### PR TITLE
Don't warn about snake case for field puns.

### LINKED ISSUE #66362: Rust shouldn't warn about snake case in patterns
For example:

```
enum Animal {
  Spider {
    iHateSnakes: bool,
  },
  ...
}

...

match pet {
  Animal::Spider { iHateSnakes } =>
  ...
```

This gives you a warning that `iHateSnakes` should be snake case in the match arm. But at that point you don't really have much choice about it. Variable name warnings should be where the variable is named, and it isn't really named there. (Maybe there's some syntax like `{ i_hate_snakes @ iHateSnakes }` but who is going to bother with that?)

This issue has been assigned to @jumbatm via [this comment](https://github.com/rust-lang/rust/issues/66362#issuecomment-557362911).

### PR DESCRIPTION
**Pull request by jumbatm**
Closes #66362.

### CODE REVIEW

**Comment by Centril** [comment]
The test case seems good.

**Comment by jumbatm** [comment]
@joshtriplett This is good for review for the actual changes now :)

**Comment by Dylan-DPC** [comment]
ferris wheel begins

r? @Centril

File: src/test/ui/lint/issue66362-no-snake-case-warning-for-field-puns.rs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#![deny(non_snake_case)]
```
Comment (Centril) [inline src/test/ui/lint/issue66362-no-snake-case-warning-for-field-puns.rs]: Rename file to `issue-66332-<the_other_stuff>.rs` (note the `-` after `issue`).

**Comment by Centril** [comment]
r=me rollup with the file renamed and the commits squashed into one.

**Comment by jumbatm** [comment]
Done

**Pull request merged**